### DISCUSSION TITLE
Question about the signature of Application::initialize

### DISCUSSION THREAD

**Discussion by amosbird** [pocoproject/poco#3638] [General]
Why do we need `Application & self` as argument instead of using `this` directly? 
 https://github.com/pocoproject/poco/blob/191cbdc97e1d9de023b8a65e0dc844911c06f6a8/Util/src/Application.cpp#L181

**Comment by aleks-f** [comment]
I suspect because it is not a virtual function

**Reply by amosbird** [reply]
Interesting. Is there any reason to not make it virtual? There are a bunch of virtual functions in the class `Application`.

**Reply by aleks-f** [reply]
It actually is virtual, just not explicitly stated in `Application`:
https://github.com/pocoproject/poco/blob/191cbdc97e1d9de023b8a65e0dc844911c06f6a8/Util/include/Poco/Util/Subsystem.h#L59

And so, it is needed because a `Subsytem` can initialize other subsystems, so `self` is not necessarily the same thing as `*this`:
https://github.com/pocoproject/poco/blob/191cbdc97e1d9de023b8a65e0dc844911c06f6a8/Util/src/Application.cpp#L186
